Pull request for supervisor review (#1)
AstroViking/ba-thesis: chapters/04-implementation.Rmd

For data points on the right side of the grid point $g_l$: The fractional parts of the data points are summed if the integral parts equal $l$. 

For data points on the left side of the grid point $g_l$: $1$ minus the fractional parts of the data points are summed if the integral parts equal $l-1$.

Including the weights this looks as follows

\begin{equation}
c_l = c(g_l) = \sum_{\substack{\tilde{x}^{frac}_k  \in \tilde{X}^{frac}\\l = \tilde{x}^{int}_k}} \tilde{x}^{frac}_k \cdot w_k + \sum_{\substack{\tilde{x}^{frac}_k  \in \tilde{X}^{frac}\\l = \tilde{x}^{int}_k + 1}} (1-\tilde{x}^{frac}_k) \cdot w_k
(\#eq:linbinnoloop)
\end{equation}

Left and right side sums can then be calculated efficiently with the TensorFlow function `tf.math.bincount`.

The binned method `ZfitBinned` is implemented in the same class definition as `ZfitExact`, the binning can be enabled by specifying a constructor argument.

## FFT based method

> The KDE method called `ZfitFFT`, which uses the FFT based method (discussed in section \@ref(fftTheory)), is implemented as a `zfit.pdf.BasePdf` class. It is not based on TensorFlow Probability as there is it does not use a `MixtureDistribution` but instead calculates the estimate for the given grid points directly. To still infer values for other points in the range of $x$ `tfp.math.interp_regular_1d_grid` is used, which computes a linear interpolation of values between the grid. In TensorFlow one-dimensional discrete convolutions are efficiently implemented already if we use `tf.nn.conv1d`. In benchmarking using this method to calculate the estimate proved significantly faster than using `tf.signal.rfft` and `tf.signal.irfft` to transform, multiply and inverse transform the vectors, which is implemented as an alternative option as well.

Reviewer: ```suggestion
The KDE method called `ZfitFFT`, which uses the FFT based method (discussed in section \@ref(fftTheory)), is implemented as a `zfit.pdf.BasePdf` class. It is not based on TensorFlow Probability as it does not use a `MixtureDistribution` but instead calculates the estimate for the given grid points directly. To still infer values for other points in the range of $x$ `tfp.math.interp_regular_1d_grid` is used, which computes a linear interpolation of values between the grid. In TensorFlow one-dimensional discrete convolutions are efficiently implemented already if we use `tf.nn.conv1d`. In benchmarking using this method to calculate the estimate proved significantly faster than using `tf.signal.rfft` and `tf.signal.irfft` to transform, multiply and inverse transform the vectors, which is implemented as an alternative option as well.
```

# Implementation {#tensorflowImplementation}

## Exact Kernel Density Estimation

> 

Reviewer: [7] I figured what we're missing here: An introduction to TensorFlow (and TensorFlow probability). Saying that it is a framework for blabla and giving an idea that a "graph" is built first.

One Paragraph is fine, but something about the framework would be great!

The implementation of an exact Kernel Density Estimation in TensorFlow is straightforward. As described in the original Tensorflow Probability Paper[@googleTFP], a KDE can be constructed by using its MixtureSameFamily Distribution, given sampled `data` as follows

```{python exactKDE, eval=FALSE}

from tensorflow_probability import distributions as tfd

f = lambda x: tfd.Independent(tfd.Normal(loc=x, scale=1.))
n = data.shape[0].value

kde = tfd.MixtureSameFamily(
    mixture_distribution=tfd.Categorical(
        probs=[1 / n] * n),
    components_distribution=f(data))
```

> Interestingly, due to the smart encapsulated structure of TensorFlow Probability we can use any distribution of the loc-scale family type as a kernel as long as it follows the Distribution contract in TensorFlow Probability. If the used Kernel has only bounded support, the implementation proposed in this paper allows to specify the support upon instantiation of the class. If the Kernel has infinite support (like a Gaussian kernel for instance) a practical support estimate is calculated by searching for approximate roots with Brent's method[@brent1971algorithm] implemented for TensorFlow in the python package `tf_quant_finance` by Google. This allows us to speed up the calculation.

Reviewer: ```suggestion
Interestingly, due to the smart encapsulated structure of TensorFlow Probability we can use any distribution of the loc-scale family type as a kernel as long as it follows the Distribution contract in TensorFlow Probability. If the used Kernel has only bounded support, the implementation proposed in this paper allows to specify the support upon instantiation of the class. If the Kernel has infinite support (like a Gaussian kernel for instance) a practical support estimate is calculated by searching for approximate roots with Brent's method[@brent1971algorithm] implemented for TensorFlow in the python package `tf_quant_finance` by Google. This allows us to speed up the calculation as negligible contributions from far away kernels are neglected.
```

Or "how" exactly is the "speeup" achieved? Write that

> However calculating an exact kernel density estimation is not always feasible as this can take a long time with a huge collection of data points, especially in high energy physics. By implementing it in TensorFlow we already get a significant speed up compared to implementations in native Python, since most of TensorFlow is actually implemented in C++ and the code is optimized before running. Even if the theoretical computational complexity remains the same.

Reviewer: ```suggestion
However calculating an exact kernel density estimation is not always feasible as this can take a long time with a huge collection of data points. By implementing it in TensorFlow we already get a significant speed up compared to implementations in native Python, since most of TensorFlow is actually implemented in C++ and the code is optimized before running. Even if the theoretical computational complexity remains the same.
```

It's mentioned quite a few times

[4] while true, it's also, most likely due to parallelization. And it's an unrealistic comparison: Numpy e.g. is run in C++ as well (or even Fortran) which would be the goto alternative.
Maybe this can somehow be brought more to the point?

And rearrange the last sentence: it feels out of place there

## Simple and linear Binning

> Simple binning is already implemented in TensorFlow in the function `tf.histogram_fixed_width`.

Reviewer: Merge with the next paragraph

> Implementing linear binning efficiently with TensorFlow is a bit tricky since loops should be avoided. However with some inspiration from the KDEpy package[@KDEpy] this can be done without using loops at all. 

Reviewer: merge as well? With loops avoided, refer to the (TODO) introduction to TensorFlow. That the graph paradigm prefers vectorized operations strongly over loops

First, every data point $x_k$  can be described by an integral part $x^{integral}_k$ (equal to its nearest left grid point number $l$ = $x^{integral}_k$) plus some fractional part $x^{fractional}_k$ (corresponding to the additional distance between grid point $g_l$ and data point $x_k$).

Then we can solve the linear binning in the following way.

> 

Reviewer: Also merge this paragraphs

For data points on the right side of the grid point $g_l$: The fractional parts of the data points are summed if the integral parts equal $l$. 

For data points on the left side of the grid point $g_l$: $1$ minus the fractional parts of the data points are summed if the integral parts equal $l-1$.

Including the weights this looks as follows

\begin{equation}
c_l = c(g_l) = \sum_{\substack{x^{fractional}_k  \in X^{fractional}\\l = x^{integral}_k}} x^{fractional}_k \cdot w_k + \sum_{\substack{x^{fractional}_k  \in X^{fractional}\\l = x^{integral}_k + 1}} (1-x^{fractional}_k) \cdot w_k
(\#eq:linbinnoloop)
\end{equation}

Left and right side sums can then be calculated efficiently with the TensorFlow function `tf.math.bincount`.

## Using convolution and the Fast Fourier Transform

> In TensorFlow one-dimensional convolutions are efficiently implemented already if we use `tf.nn.conv1d`. In benchmarking using this method proved significantly faster than using `tf.signal.rfft` and `tf.signal.irfft` to transform, multiply and inverse transform the vectors, which is implemented as an alternative as well.

Reviewer: [6] "... one-dimensional DISCRETE convolutions..."

This algorithm is implemented as its own class since it does not represent a complete mixture distribution anymore but calculates just the density distribution values at the specified grid points. To still infer values for other points in the range of $x$ `tfp.math.interp_regular_1d_grid` is used, which computes a linear interpolation of values between the grid.

## Improved Sheather Jones Algorithm

> To find the roots for equation \@ref(eq:hamisegamma2) Brent's method[@brent1971algorithm] was used, which is implemented in TensorFlow in the python package `tf_quant_finance`. For the Discrete Cosine Transform `tf.signal.dct` is used.

Reviewer: [4] Maybe extend a little bit. Was there any other complication? Or something to sketch for the implementation?

> To find the roots for equation \@ref(eq:hamisegamma2) Brent's method[@brent1971algorithm] was used, which is implemented in TensorFlow in the python package `tf_quant_finance`. For the Discrete Cosine Transform `tf.signal.dct` is used.

Reviewer: You already mentioned the method `tf_quant_finance` before for that method

## Source Code

> 

Reviewer: Move this to the appendix

However calculating an exact kernel density estimation is not always feasible as this can take a long time with a huge collection of data points. By implementing it in TensorFlow we already get a significant speed up compared to implementations in native Python, due to TensorFlow's advantages mentioned above. Nonetheless the computational complexity remains the same and for large data this can still make the exact KDE impractical.

> An exact kernel density estimation using zfit called `ZfitExact` is implemented as a `zfit.pdf.WrapDistribution` class, which is zfit's class type for wrapping TensorFlow Probability distributions.

Reviewer: ```suggestion
An exact KDE using zfit called `ZfitExact` is implemented as a `zfit.pdf.WrapDistribution` class, which is zfits class type for wrapping TensorFlow Probability distributions.
```
